File: backend/services/logs.js
const { exec } = require("child_process");
const { promisify } = require("util");
const fs = require("fs");
const path = require("path");

const execPromise = promisify(exec);

// Get available log sources
async function getLogSources() {
    const sources = [
        { id: "panel", name: "Panel Application", type: "file" },
        { id: "docker", name: "Docker Containers", type: "docker", available: false },
        { id: "pm2", name: "PM2 Processes", type: "pm2", available: false }
    ];

    // Check Docker availability
    try {
        const dockerService = require("./docker");
        const dockerStatus = await dockerService.checkDockerAvailable();
        const dockerAvailable = dockerStatus.available;
        const dockerSource = sources.find(s => s.id === "docker");
        if (dockerSource) dockerSource.available = dockerAvailable;
    } catch (err) {
        // Docker not available
    }

    // Check PM2 availability
    try {
        const pm2Service = require("./pm2");
        const pm2Status = await pm2Service.checkPm2Available();
        const pm2Available = pm2Status.available;
        const pm2Source = sources.find(s => s.id === "pm2");
        if (pm2Source) pm2Source.available = pm2Available;
    } catch (err) {
        // PM2 not available
    }

    return sources;
}

// Read panel logs
async function readPanelLogs(lines = 100) {
    const logFile = path.join(__dirname, "../../logs/panel.log");

    try {
        if (!fs.existsSync(logFile)) {
            return "No logs available yet";
        }

        if (process.platform === "win32") {
            const { stdout } = await execPromise(`powershell -Command "Get-Content '${logFile}' -Tail ${lines}"`);
            return stdout;
        } else {
            const { stdout } = await execPromise(`tail -n ${lines} "${logFile}"`);
            return stdout;
        }
    } catch (error) {
        return `Error reading logs: ${error.message}`;
    }
}

// Read Docker container logs
async function readDockerLogs(containerName, lines = 100) {
    try {
        const dockerService = require("./docker");
        const containers = await dockerService.listContainers();
        const container = containers.find(c => c.name === containerName);

        if (!container) {
            return `Container "${containerName}" not found`;
        }

        const logs = await dockerService.getContainerLogs(container.id, lines);
        return logs;
    } catch (error) {
        return `Error reading Docker logs: ${error.message}`;
    }
}

// Read PM2 process logs
async function readPm2Logs(processName, lines = 100) {
    try {
        const pm2Service = require("./pm2");
        const logs = await pm2Service.getProcessLogs(processName, lines);
        return logs;
    } catch (error) {
        return `Error reading PM2 logs: ${error.message}`;
    }
}

// Get logs from source
async function getLogsFromSource(sourceId, target, lines = 100) {
    switch (sourceId) {
        case "panel":
            return await readPanelLogs(lines);

        case "docker":
            if (!target) {
                return "Please specify a container name";
            }
            return await readDockerLogs(target, lines);

        case "pm2":
            if (!target) {
                return "Please specify a process name";
            }
            return await readPm2Logs(target, lines);

        default:
            return "Unknown log source";
    }
}

// Get targets for a source (e.g., list of containers or processes)
async function getLogTargets(sourceId) {
    switch (sourceId) {
        case "docker":
            try {
                const dockerService = require("./docker");
                const containers = await dockerService.listContainers();
                return containers.map(c => ({ id: c.name, name: c.name }));
            } catch (err) {
                return [];
            }

        case "pm2":
            try {
                const pm2Service = require("./pm2");
                const processes = await pm2Service.listProcesses();
                return processes.map(p => ({ id: p.name, name: p.name }));
            } catch (err) {
                return [];
            }

        default:
            return [];
    }
}

// Search in logs
function searchInLogs(logs, query) {
    if (!query) return logs;

    const lines = logs.split('\n');
    const filtered = lines.filter(line =>
        line.toLowerCase().includes(query.toLowerCase())
    );

    return filtered.join('\n');
}

module.exports = {
    getLogSources,
    getLogsFromSource,
    getLogTargets,
    searchInLogs
};

File: be/internal/logs/logs.go
package logs

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

type Service struct {
	root string
}

func New(root string) *Service { return &Service{root: root} }

type Source struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Type      string `json:"type"`
	Available bool   `json:"available,omitempty"`
}

type Target struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

func (s *Service) Sources(ctx context.Context) []Source {
	return []Source{
		{ID: "panel", Name: "Panel Application", Type: "file"},
		{ID: "docker", Name: "Docker Containers", Type: "docker", Available: commandAvailable(ctx, "docker", "version")},
		{ID: "pm2", Name: "PM2 Processes", Type: "pm2", Available: commandAvailable(ctx, "pm2", "-v")},
	}
}

func (s *Service) Logs(ctx context.Context, sourceID, target string, lines int) string {
	lines = clampLines(lines)
	switch sourceID {
	case "panel":
		return s.panelLogs(lines)
	case "docker":
		if target == "" {
			return "Please specify a container name"
		}
		return dockerLogs(ctx, target, lines)
	case "pm2":
		if target == "" {
			return "Please specify a process name"
		}
		return pm2Logs(ctx, target, lines)
	default:
		return "Unknown log source"
	}
}

func (s *Service) Targets(ctx context.Context, sourceID string) []Target {
	switch sourceID {
	case "docker":
		return dockerTargets(ctx)
	case "pm2":
		return pm2Targets(ctx)
	default:
		return []Target{}
	}
}

func Search(logs, query string) string {
	if query == "" {
		return logs
	}
	q := strings.ToLower(query)
	var out []string
	for _, line := range strings.Split(logs, "\n") {
		if strings.Contains(strings.ToLower(line), q) {
			out = append(out, line)
		}
	}
	return strings.Join(out, "\n")
}

func (s *Service) panelLogs(lines int) string {
	p := filepath.Join(s.root, "logs", "panel.log")
	b, err := os.ReadFile(p)
	if os.IsNotExist(err) {
		return "No logs available yet"
	}
	if err != nil {
		return "Error reading logs: " + err.Error()
	}
	return lastLines(string(b), lines)
}

func dockerLogs(ctx context.Context, containerName string, lines int) string {
	if !targetExists(ctx, "docker", containerName) {
		return fmt.Sprintf("Container %q not found", containerName)
	}
	out, err := run(ctx, "docker", "logs", "--tail", fmt.Sprint(lines), containerName)
	if err != nil {
		return "Error reading Docker logs: " + err.Error()
	}
	return out
}

func pm2Logs(ctx context.Context, processName string, lines int) string {
	out, err := run(ctx, "pm2", "logs", processName, "--lines", fmt.Sprint(lines), "--nostream", "--raw")
	if err != nil {
		return "Error reading PM2 logs: " + err.Error()
	}
	return out
}

func dockerTargets(ctx context.Context) []Target {
	out, err := run(ctx, "docker", "ps", "-a", "--format", "{{.Names}}")
	if err != nil {
		return []Target{}
	}
	var targets []Target
	for _, line := range strings.Split(strings.TrimSpace(out), "\n") {
		name := strings.TrimSpace(line)
		if name != "" {
			targets = append(targets, Target{ID: name, Name: name})
		}
	}
	return targets
}

func pm2Targets(ctx context.Context) []Target {
	out, err := run(ctx, "pm2", "jlist")
	if err != nil {
		return []Target{}
	}
	var processes []struct {
		Name string `json:"name"`
	}
	if json.Unmarshal([]byte(out), &processes) != nil {
		return []Target{}
	}
	targets := make([]Target, 0, len(processes))
	for _, p := range processes {
		if p.Name != "" {
			targets = append(targets, Target{ID: p.Name, Name: p.Name})
		}
	}
	return targets
}

func targetExists(ctx context.Context, source, target string) bool {
	for _, t := range dockerTargets(ctx) {
		if source == "docker" && t.Name == target {
			return true
		}
	}
	return false
}

func commandAvailable(ctx context.Context, name string, args ...string) bool {
	if _, err := exec.LookPath(name); err != nil {
		return false
	}
	_, err := run(ctx, name, args...)
	return err == nil
}

func run(ctx context.Context, name string, args ...string) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, 8*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, name, args...)
	if runtime.GOOS != "windows" {
		cmd.Env = os.Environ()
	}
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &out
	if err := cmd.Run(); err != nil {
		return out.String(), err
	}
	return out.String(), nil
}

func clampLines(lines int) int {
	if lines <= 0 {
		return 100
	}
	if lines > 5000 {
		return 5000
	}
	return lines
}

func lastLines(s string, n int) string {
	parts := strings.Split(strings.ReplaceAll(s, "\r\n", "\n"), "\n")
	if len(parts) > 0 && parts[len(parts)-1] == "" {
		parts = parts[:len(parts)-1]
	}
	if len(parts) > n {
		parts = parts[len(parts)-n:]
	}
	return strings.Join(parts, "\n")
}
